conda/env/env.py

```diff
@@ -207,7 +222,430 @@ def add(self, package_name):
+    For example:
+
+    {
+        "if": "__linux",    # <-- if __linux is present, also include flask as a requirement
+        "then": "flask",
+    }
+    """
+
+    def __init__(self, spec: str | dict[str, str]):
+        if isinstance(spec, str):
+            self.spec = spec
+            self.condition = None
+        elif isinstance(spec, dict):
+            for key in ["if", "then"]:
+                if key not in spec:
+                    raise ValueError(f"Conditional requirement missing an `{key}` key.")
+
+            self.spec = spec["then"]
+            self.condition = spec["if"]
+        else:
+            raise ValueError(f"Invalid requirement: {spec}")
+
+    def __repr__(self) -> str:
+        result = self.spec
+        if self.condition:
+            result = f"{result} (if {self.condition})"
+        return result
+
+    def __eq__(self, other: Any) -> bool:
+        if isinstance(other, Requirement):
+            return self.spec == other.spec
+
+        raise NotImplementedError
+
+    def __lt__(self, other: Requirement) -> bool:
+        return self.spec < other.spec
+
+    def serialize(self) -> str | dict[str, str]:
+        """Serialize the package requirement.
+
+        Can be a conditional requirement of the form
+
+        {
+            "if": "__win",
+            "then": "flask"
+        }
+
+        :return: Serialized Requirement
+        """
+        if self.condition:
+            return {"if": self.condition, "then": self.spec}
+        return self.spec
+
+    def __str__(self) -> str:
+        return repr(self)
+
+
+class PypiRequirement(Requirement):
+    pass
+
+
+class CondaRequirement(Requirement):
+    pass
+
+
+StringConditionalRequirement = dict[str, str]
+StringRequirements = Iterable[Union[str, StringConditionalRequirement]]
+
+
+class EnvironmentConfig:
+    """A class that contains configuration variables for an environment.
+
+    These variables override user configuration, e.g. settings in .condarc.
+    """
+
+    def __init__(self, **options):
+        self.options = options
+
+    def __repr__(self) -> str:
+        lines = []
+        for key, value in self.options.items():
+            lines.append(f"{key}: {value}")
+        return "\n".join(lines)
+
+    def serialize(self) -> dict[str, str]:
+        """Serialize the environment configuration.
+
+        Since this is a thin wrapper around a dictionary, just return the underlying
+        dict.
+
+        :return: A dict of primitives containing configuration options for the
+            environment
+        """
+        return self.options
+
+
+class EnvironmentBase:
+    """A base class representing an ``environment.yaml`` file"""
+
+    def to_dict(self) -> dict:
+        raise NotImplementedError
+
+    def to_yaml(self) -> str | None:
+        raise NotImplementedError
+
+    def save(self):
+        raise NotImplementedError
+
+
+class EnvironmentV2(EnvironmentBase):
```

Comment:
Should this be implemented as a EnvSpec eg. in `conda.env.specs` instead of as a new type of environment?

Reply:
That would be fine by me!

Reply:
Just want to clarify my question above. I'm getting a bit confused between what ought to be a user facing environement spec and an internal representation of an environment. Currently, the `conda.env.env.Environment` class is pulling double duty and representing both of these ideas.

Thinking about this more, I think it's a good idea to seperate these two ideas. Since the environment spec plugin stuff has been merged, I think the user facing (environment file definition) part of this can be implemented as a plugin.